### PR TITLE
Api provides admin functionality to add an image

### PR DESCRIPTION
**Pull request by offcast-dev**
PT: https://www.pivotaltracker.com/story/show/178092234
### User Story:
```
As an Api
In order to have the client show specific images for every article
I need to be able to attach an image to an Article
```
___________
### Changes Proposed:
- adds image attachment to article
- sets up base64 decoding
- adds image to article factory
- sets up aws s3 storage 
- seeds articles with unique images
___________
### Packages / Gems Added:
- aws-sdk-s3
___________
### What we Learned:
- how to connect aws storage
- how to attach files
- how to work with base64 encoding
- some mistakes do not appear in automated tests

### CODE REVIEW

File: app/controllers/api/articles_controller.rb [hunk 1]
```diff
@@ -42,10 +42,14 @@ def update
 
   private
 
+  def attach_image(article)
+    params[:article][:image].present? && DecodeService.attach_image(params[:article][:image].first, article.image)
```
Comment (LucasKnudsen) [inline app/controllers/api/articles_controller.rb]: We are sending the image as a string (not an array), so the `.first` will fack it up :P

File: spec/requests/api/journalist/journalist_can_create_articles_spec.rb [hunk 2]
```diff
@@ -1,6 +1,9 @@
 RSpec.describe 'POST /api/articles', type: :request do
   let(:journalist) { create(:user, role: 'journalist') }
   let(:auth_headers) { journalist.create_new_auth_token }
+  let(:image) do
+    [ File.read(fixture_path + '/fake-news-fixture.txt') ]
```
Comment (LucasKnudsen) [inline spec/requests/api/journalist/journalist_can_create_articles_spec.rb]: Same here with the array, can remove

File: app/serializers/articles_index_serializer.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 class ArticlesIndexSerializer < ActiveModel::Serializer
-  attributes :id, :title, :teaser, :date, :category
+  attributes :id, :title, :teaser, :date, :category, :image
```
Comment (LucasKnudsen) [inline app/serializers/articles_index_serializer.rb:2]: Are we testing this change? What's the value of image?

File: app/serializers/articles_show_serializer.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 class ArticlesShowSerializer < ActiveModel::Serializer
-  attributes :id, :title, :teaser, :body, :date, :author, :category
+  attributes :id, :title, :teaser, :body, :date, :author, :category, :image
```
Comment (LucasKnudsen) [inline app/serializers/articles_show_serializer.rb:2]: And this? Are we sure it's an url?

File: app/controllers/api/articles_controller.rb (hunk 1 above)
Comment (offcast-dev) [reply app/controllers/api/articles_controller.rb]: ok, fixed this

File: spec/requests/api/journalist/journalist_can_create_articles_spec.rb (hunk 2 above)
Comment (offcast-dev) [reply spec/requests/api/journalist/journalist_can_create_articles_spec.rb]: and this

File: app/serializers/articles_index_serializer.rb (hunk 3 above)
Comment (offcast-dev) [reply app/serializers/articles_index_serializer.rb:2]: I am not sure about it to be honest. I think image will be an object. In test environment if we try to get to url in serializer (like in PT), all tests fail as it cannot get attribute url of image, but in Rails console url is accessible in article.image.url. I am not really sure what to think about it, probably need to spin up server, create an article, than get one, and look at response. Can do it after dinner

File: app/serializers/articles_show_serializer.rb (hunk 4 above)
Comment (LucasKnudsen) [reply app/serializers/articles_show_serializer.rb:2]:
```rb
def image  
  Rails.env.test? ?
    ActiveStorage::Blob.service.path_for(image.key) 
    : 
    image.service_url(expires_in: 1.hour, disposition: 'inline') 
end
```

Comment (LucasKnudsen) [reply app/serializers/articles_show_serializer.rb:2]: Perhaps we need something like this

Comment (christerforsgren91) [reply app/serializers/articles_show_serializer.rb:2]: what goin on here?

File: spec/requests/api/journalist/journalist_can_create_articles_spec.rb (hunk 2 above)
Comment (offcast-dev) [reply spec/requests/api/journalist/journalist_can_create_articles_spec.rb]: Nah.. doesnt work as a string. Gives this error ActiveSupport::MessageVerifier::InvalidSignature when we create article.
You will have to send an array to us

**Review by lukeperrin10 — APPROVED** [review]

File: app/controllers/api/articles_controller.rb [hunk 5]
```diff
@@ -43,13 +43,13 @@ def update
   private
 
   def attach_image(article)
-    params[:article][:image].present? && DecodeService.attach_image(params[:article][:image].first, article.image)
+    params[:article][:image].present? && DecodeService.attach_image(params[:article][:image], article.image)
```
Comment (LucasKnudsen) [inline app/controllers/api/articles_controller.rb:46]: So it's working now without the array? :D

File: app/serializers/articles_show_serializer.rb [hunk 6]
```diff
@@ -11,4 +11,12 @@ def author
   def date
     object.created_at.strftime('%F, %H:%M')
   end
+
+  def image
```
Comment (LucasKnudsen) [inline app/serializers/articles_show_serializer.rb:15]: Awesome, glad `image` works!

File: db/seeds.rb [hunk 7]
```diff
@@ -39,3 +51,9 @@
   Article.create(title: titles[i], teaser: teasers[i], body: body[i], category: categories[rand(categories.count - 1)],
                  user_id: user.id)
 end
+
+Article.all.each_with_index do |article, index|
+  file = URI.open(image_url[index])
+  
+  article.image.attach(io: file, filename: 'article_image.jpg', content_type: 'image/jpg')
+end
```
Comment (LucasKnudsen) [inline db/seeds.rb:59]: 👍

**Review by LucasKnudsen — COMMENTED** [review]
My only question would be that is the `image` attribute tested for the other requests?

File: app/serializers/articles_show_serializer.rb (hunk 4 above)
Comment (offcast-dev) [reply app/serializers/articles_show_serializer.rb:2]: This Amazon remote storage works in development environment but not in test environment. So when we run tests we don't have image url, so the line ```image.service_url(expires_in: 1.hour, disposition: 'inline') ``` errors out. We can check what environment we are in and make serializer return different stuff in different environments. So the tests are green and we get our urls when deployed

File: app/controllers/api/articles_controller.rb (hunk 5 above)
Comment (offcast-dev) [reply app/controllers/api/articles_controller.rb:46]: yes, as soon as deployed we will see images on frontend, checked on localhost.

**Review by lukeperrin10 — APPROVED** [review]

File: app/serializers/articles_show_serializer.rb (hunk 4 above)
Comment (LucasKnudsen) [reply app/serializers/articles_show_serializer.rb:2]: Awesum stuff

**Review by LucasKnudsen — APPROVED** [review]

**Pull request merged**